### PR TITLE
Fix trigger mcontrol.chain match issue #599 #627

### PR DESCRIPTION
**Pull request by YenHaoChen**
A false condition of variable chain_ok is missing. The false condition should be mcontrol.chain=1 and not matching; otherwise, the chain_ok=true (including initialization).

Related issues:
https://github.com/riscv-software-src/riscv-isa-sim/issues/627
https://github.com/riscv-software-src/riscv-isa-sim/issues/599

### CODE REVIEW

**Comment by scottj97** [comment]
I don't really understand the logic here but I do observe a useful change in triggers.

I have a chained breakpoint set up like so:
tdata1[0]=0x200000000000090b # chain=1 match=2(ge) u=1 store=1 load=1
tdata2[0]=0x000000008000300a
tdata1[1]=0x200000000000018b # match=3(lt) u=1 store=1 load=1
tdata2[1]=0x0000000080003013

This represents a breakpoint on 0x8000300a <= addr < 0x80003013.

Now execute a `ld` to 0x80003002, which loads bytes up through 0x80003009.

Before this PR, this `ld` would trigger a breakpoint trap. Why? It wasn't touching any of the bytes in the range.

After this PR, no breakpoint trap is taken on this instruction.

Note that for aligned accesses, Spike isn't checking every byte in the range -- only the first. For misaligned, it checks every byte. For example, `ld` to 0x80003008 does not trigger the above breakpoint following this PR. (The debug spec under `mcontrol.select` recommends but does not require checking every byte.)

**Comment by YenHaoChen** [comment]
*Update commit message.

```
The variable chain_ok is used to indicate if the current trigger
is suppressed by the trigger chain. A true value means the trigger
is either un-chained or matches all previous triggers in the chain,
and a false value means the trigger is chained and mismatches
previous triggers.

A false condition of variable chain_ok is missing. The false
condition should be mcontrol.chain=1 and not matching; otherwise,
the chain_ok=true (including initialization). The bug results in
issues https://github.com/riscv-software-src/riscv-isa-sim/pull/559 and https://github.com/riscv-software-src/riscv-isa-sim/issues/627.

Related issues:
- https://github.com/riscv-software-src/riscv-isa-sim/issues/599
- https://github.com/riscv-software-src/riscv-isa-sim/issues/627

This PR fixes the issues https://github.com/riscv-software-src/riscv-isa-sim/pull/559 and https://github.com/riscv-software-src/riscv-isa-sim/issues/627.
```

**Comment by YenHaoChen** [comment]
@timsifive @scottj97 
Is there anything missing from my side?

File: riscv/triggers.cc [hunk 1]
```diff
@@ -183,7 +183,7 @@ match_result_t module_t::memory_access_match(action_t * const action, operation_
       return result;
     }
 
-    chain_ok = true;
+    chain_ok = (result!=MATCH_NONE || !triggers[i]->chain());
```
Comment (timsifive) [inline riscv/triggers.cc]:
```suggestion
    chain_ok = result != MATCH_NONE || !triggers[i]->chain();
```

**Review by timsifive — CHANGES_REQUESTED** [review]
The logic looks right. Did you check that https://github.com/riscv-software-src/riscv-tests/tree/master/debug tests still pass?

**Comment by YenHaoChen** [comment]
> The logic looks right. Did you check that https://github.com/riscv-software-src/riscv-tests/tree/master/debug tests still pass?

I have checked that riscv-tests still passes with ```make all```.

**Review by timsifive — APPROVED** [review]

**Pull request merged**